src/codeu/chat/server/LocalFile.java

```diff
@@ -0,0 +1,166 @@
+import codeu.chat.common.User;
+import codeu.chat.util.Logger;
+import codeu.chat.util.Uuid;
+
+public class LocalFile implements Serializable
+{
+    private static final long serialVersionUID = 1L;
+    //Instance varibles for saving the current data of server.
+    private final HashMap<Uuid,User> users;
+    private final HashMap<Uuid,ConversationHeader> conversationHeaders;
+    private final HashMap<Uuid,Message> messages;
+
+    private transient final File file;
+
+    private static final Logger.Log LOG = Logger.newLog(LocalFile.class);
+
+    private transient boolean hasModified = false;//It indicates if there is a new data should be handled.
+    public LocalFile(File file)
+    {
+        this.file = file;
+        LocalFile newClass;
+        HashMap<Uuid,User>tempUsers = new HashMap<>();
+        HashMap<Uuid,ConversationHeader>tempConversationHeaders= new HashMap<>();
+        HashMap<Uuid,Message>tempMessages= new HashMap<>();
+        try
+        {
+            //Deserialization
+            RandomAccessFile randomAccessFile = new RandomAccessFile(file, "r");
+            FileInputStream fileInputStream = new FileInputStream(randomAccessFile.getFD());
+		    ObjectInputStream objectInputStream = new ObjectInputStream(fileInputStream);
+            //Read the information to current instance
+		    newClass = (LocalFile)objectInputStream.readObject();
+            tempUsers = newClass.users;
+            tempConversationHeaders = newClass.conversationHeaders;
+            tempMessages = newClass.messages;
+
+            fileInputStream.close();
+            objectInputStream.close();
+        }
+        catch(Exception exception)
+        {
+            System.out.println("ERROR:Unable to read data.");
+            System.out.println("WARNING:If it is the case that the file doesn't exist, a new local file will be created.");
+            exception.printStackTrace();
+        }
+        finally
+        {
+            //Write temporary information to the instance varibles
+            users = tempUsers;
+            conversationHeaders = tempConversationHeaders;
+            messages = tempMessages;
+        }
+    }
+    /**
+     * Save the data into local file.
+     * 
+     * @exception   Failed to save data.
+     */
+    public void saveData() throws IOException
+    {
+        if(!hasModified)//Only save the data when the data is updated.
+        {
+            return;
+        }
+        hasModified = false;
+        if(!file.exists())
+        {
+            file.createNewFile();
+        }
+        //Serialization
+        RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");
+        FileOutputStream fileOutputStream = new FileOutputStream(randomAccessFile.getFD());
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream);
+        objectOutputStream.writeObject(this); 
+
+        LOG.info("Data saved!");
+
+        fileOutputStream.close();
```

Comment:
If there's an exception in this method, these lines won't be executed, and these streams won't ever be closed. This is a memory leak. A nice way of avoiding this is to use a ["try-with-resources"](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) block, which is a relatively new Java feature. It will automatically close these streams when you're done with them, even if there's an exception. Here's what it would look like: 

```java
try (FileOutputStream fileOutputStream = new FileOutputStream(randomAccessFile.getFD());
     ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream) ) {

       ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream);
       objectOutputStream.writeObject(this); 
       LOG.info("Data saved!");
}
```